lib/rmg/TreasurePlacer.cpp

```diff
@@ -0,0 +1,834 @@
+				
+				this->questArtZone->getModificator<TreasurePlacer>()->possibleObjects.push_back(generateArtInfo(artid));
+				
+				return obj;
+			};
+			
+			possibleObjects.push_back(oi);
+			
+			oi.generateObject = [i, randomAppearance, this, generateArtInfo]() -> CGObjectInstance *
+			{
+				auto factory = VLC->objtypeh->getHandlerFor(Obj::SEER_HUT, randomAppearance);
+				auto obj = (CGSeerHut *) factory->create(ObjectTemplate());
+				obj->rewardType = CGSeerHut::RESOURCES;
+				obj->rID = Res::GOLD;
+				obj->rVal = generator.getConfig().questRewardValues[i];
+				
+				obj->quest->missionType = CQuest::MISSION_ART;
+				ArtifactID artid = *RandomGeneratorUtil::nextItem(generator.getQuestArtsRemaning(), generator.rand);
+				obj->quest->m5arts.push_back(artid);
+				obj->quest->lastDay = -1;
+				obj->quest->isCustomFirst = obj->quest->isCustomNext = obj->quest->isCustomComplete = false;
+				
+				generator.banQuestArt(artid);
+				
+				this->questArtZone->getModificator<TreasurePlacer>()->possibleObjects.push_back(generateArtInfo(artid));
+				
+				return obj;
+			};
+			
+			possibleObjects.push_back(oi);
+		}
+	}
+}
+
+bool TreasurePlacer::isGuardNeededForTreasure(int value)
+{
+	return zone.getType() != ETemplateZoneType::WATER && value > minGuardedValue;
+}
+
+std::vector<ObjectInfo> TreasurePlacer::prepareTreasurePile(const CTreasureInfo& treasureInfo)
+{
+	std::vector<ObjectInfo> objectInfos;
+	int maxValue = treasureInfo.max;
+	int minValue = treasureInfo.min;
+	
+	const ui32 desiredValue = generator.rand.nextInt(minValue, maxValue);
+	
+	int currentValue = 0;
+	bool hasLargeObject = false;
+	while(currentValue <= (int)desiredValue - 100) //no objects with value below 100 are available
+	{
+		auto oi = getRandomObject(desiredValue, currentValue, !hasLargeObject);
+		if(oi.value == 0) //fail
+			break;
+		
+		if(oi.templ.isVisitableFromTop())
+		{
+			objectInfos.push_back(oi);
+		}
+		else
+		{
+			if(hasLargeObject)
+			{
+				//only one large object per treasure pile
+				continue;
+			}
+			else
+			{
+				objectInfos.insert(objectInfos.begin(), oi); //large object shall at first place
+				hasLargeObject = true;
+			}
+		}
+		
+		//remove from possible objects
+		auto oiptr = std::find(possibleObjects.begin(), possibleObjects.end(), oi);
+		assert(oiptr->maxPerZone > 0);
+		oiptr->maxPerZone--;
+		
+		currentValue += oi.value;
+	}
+	
+	return objectInfos;
+}
+
+rmg::Object TreasurePlacer::constuctTreasurePile(const CTreasureInfo& treasureInfo)
+{
+	std::vector<ObjectInfo> objectInfos, tempObjectInfos;
+	rmg::Object rmgTreasurePile;
+	int maxValue = treasureInfo.max;
+	int minValue = treasureInfo.min;
+	
+	const ui32 desiredValue = generator.rand.nextInt(minValue, maxValue);
+	
+	int currentValue = 0;
+	bool hasLargeObject = false;
+	while(currentValue <= (int)desiredValue - 100) //no objects with value below 100 are available
+	{
+		auto oi = getRandomObject(desiredValue, currentValue, !hasLargeObject);
```

Comment:
Not exactly. Desired value is the **minimal** value of a pile. The max value of a pile remains constant.

Otherwise piles with values, let's say 5000-10000 would almost never generate objects with valueexactly 10000.

Reply:
Thanks! Looks like the bug was presented in "old" rmg as well. Fixed